project/apps/web/app/questions/components/question-table/QuestionTable.tsx

```diff
@@ -0,0 +1,154 @@
+} from "@/components/data-table/DataTable";
+
+import { columns } from "./columns";
+import { useQuestionsState } from "@/contexts/QuestionsStateContext";
+import { QuestionTableToolbar } from "./QuestionTableToolbar";
+import {
+  ColumnFiltersState,
+  PaginationState,
+  SortingState,
+  Updater,
+} from "@tanstack/react-table";
+import { startTransition, useState } from "react";
+import {
+  CATEGORY,
+  COMPLEXITY,
+} from "@repo/dtos/generated/enums/questions.enums";
+import useDebounce from "@/hooks/useDebounce";
+
+export function QuestionTable() {
+  const { confirmLoading, setConfirmLoading } = useQuestionsState();
+
+  const [pagination, setPagination] = useState<PaginationState>({
+    pageIndex: 0,
+    pageSize: 10,
+  });
+
+  const resetPagination = () => {
+    setPagination((prev) => ({
+      ...prev,
+      pageIndex: 0,
+    }));
+  };
+
+  const [sorting, setSorting] = useState<SortingState>([
+    {
+      id: "q_title",
+      desc: false,
+    },
+  ]);
+
+  const [columnFilters, setColumnFilters] = useState<ColumnFiltersState>([]);
+  const debouncedColumnFilters = useDebounce(
+    columnFilters,
+    100,
+    () => setConfirmLoading(true),
+    () => setConfirmLoading(false),
+  );
+
+  const { data } = useSuspenseQuery<QuestionCollectionDto>({
+    queryKey: [
+      QUERY_KEYS.Question,
+      pagination,
+      sorting,
+      debouncedColumnFilters,
+    ],
+    queryFn: async () => {
+      setConfirmLoading(true);
+      try {
+        const title = debouncedColumnFilters.find((f) => f.id === "q_title")
+          ?.value as string;
+
+        const categories = debouncedColumnFilters.find(
+          (f) => f.id === "q_category",
+        )?.value as CATEGORY[];
+
+        const complexities = debouncedColumnFilters.find(
+          (f) => f.id === "q_complexity",
+        )?.value as COMPLEXITY[];
+
+        const offset = pagination.pageIndex * pagination.pageSize;
+        const limit = pagination.pageSize;
+
+        const sort = sorting.map(
+          (s) =>
+            ({
+              field: s.id,
+              order: s.desc ? "desc" : "asc",
+            }) as SortQuestionsQueryDto,
+        );
+
+        const queryParams: QuestionFiltersDto = {
+          title,
+          categories,
+          complexities,
+          offset,
+          limit,
+          sort,
+        };
+
+        return await fetchQuestions(queryParams);
+      } finally {
+        setConfirmLoading(false);
+      }
+    },
+  });
+
+  const onPaginationChange = (updater: Updater<PaginationState>) => {
+    startTransition(() => {
+      setPagination(updater);
+    });
+  };
+
+  const onSortingChange = (updater: Updater<SortingState>) => {
+    startTransition(() => {
+      setSorting(updater);
+      resetPagination();
+    });
+  };
+
+  const onColumnFiltersChange = (updater: Updater<ColumnFiltersState>) => {
+    startTransition(() => {
+      setColumnFilters(updater);
+      resetPagination();
+    });
+  };
```

Comment:
perhaps all of these can take startTransition from `const [isPending, startTransition] = useTransition()` to provide the loading screen when any of these happens. Also probably dont need a startTransition for the last one right since the query is dependent on `debouncedcolumnfilters` rather than `columnFilters`, then maybe reset pagination should be called for debounced one too

Reply:
I included startTransition in onColumnFiltersChange to maintain consistency across all on...Change functions, and it’s also necessary for resetPagination().

The loading screen is displayed whenever a query is triggered (refer to the comment below).

The useDebounce hook is needed to trigger the loading screen as soon as the user starts typing, rather than only when the query is sent, for a better UX.

Reply:
ah i was thinking for example if i select and unselect a column then because its debounced, nothing should happen but in this case we would have reset the pagination alr

Reply:
ah I see your point. at the moment i didn't include any debouncing for pagination and sorting - so web queries immediately on button clicks.